File: Difficulty: Medium/Bottom View of Binary Tree/bottom-view-of-binary-tree.cpp
/*
struct Node
{
    int data;
    Node* left;
    Node* right;
};
*/

#define vpii vector<pair<int,int>>

class Solution {
  public:
     // Function to return a list containing the inorder traversal of the tree.
    void dfs(Node* root,int level, int cord,vpii& pos, vpii& neg ) {
        if(root==nullptr) return;
        dfs(root->left,level+1,cord-1,pos,neg);
        
        if(cord>=0) {
            if(pos.size()<=cord) pos.resize(cord+1,{-1,-1});
            if(pos[cord].first<=level) {
                pos[cord].first = level;
                pos[cord].second = root->data;
            }
            // else if(pos.size()<=cord) {
            //     pos.push_back({level,root->data});
            // }
        }
        else {
            int negCord = (0-cord);
             if(neg.size()<=negCord) neg.resize(negCord+1,{-1,-1});
            if(neg[negCord].first<=level) {
                neg[negCord].first = level;
                neg[negCord].second = root->data;
            }
            // else if(neg.size()<=cord) {
            //     neg.push_back({level,root->data});
            // }
        }
        
        dfs(root->right,level+1,cord+1,pos,neg);
    }
    vector<int> bottomView(Node *root) {
        // Your Code Here
        vector<int> ans;
        vpii pos,neg;
        //neg.push_back({-1,-1});
        
     
        int level = 0, cord = 0;
        dfs(root,level,cord,pos,neg);
        
        for(int i = neg.size()-1;i>0;i--) {
            ans.push_back(neg[i].second);
        }
        for(int i = 0;i<pos.size();i++) {
            ans.push_back(pos[i].second);
        }
        
        
        return ans;
    }
};

File: Difficulty: Medium/Bottom View of Binary Tree/bottom-view-of-binary-tree.java
/*
Definition for Node
class Node {
    int data;
    Node left;
    Node right;

    Node(int val) {
        data = val;
        left = right = null;

    }
}
*/


class Pair<T,P> {
    private T first;
    private P second;
    private Pair(T f, P s) {
        this.first = f;
        this.second = s;
    }

    public T first() {
        return this.first;
    }

    public P second() {
        return this.second;
    }

    public void first(T f) {
        this.first = f;
    }

    public void second(P s) {
        this.second = s;
    }

    public  static <T,P> Pair<T,P> makePair(T a, P b) {
        return new Pair<T,P>(a,b);
    }


}

class Solution {
    
    private void solve(Node root, int level, int dir, TreeMap<Integer,Pair<Integer,Integer>> ans) {
        if(root==null) return;
        if(ans.containsKey(dir)) {
            if(ans.get(dir).first()<=level) {
                ans.get(dir).first(level);
                ans.get(dir).second(root.data);
            }
        }
        else {
            //ans.put(dir,new Pair<Integer,Integer>(level,root.data));
            ans.put(dir,Pair.makePair(level,root.data));
        }
        solve(root.left,level+1,dir-1,ans);
        solve(root.right,level+1,dir+1,ans);

    }
    
    public ArrayList<Integer> bottomView(Node root) {
        // code here
        
         TreeMap<Integer,Pair<Integer,Integer>> ans = new TreeMap<>();
         ArrayList<Integer> ans2 = new ArrayList<>();
         solve(root,0,0,ans);
        // System.out.println(ans);
         for(int a: ans.keySet()) {
             //System.out.print(ans.get(a)+" ");
             ans2.add(ans.get(a).second());
         }
         return ans2;
    }
}